draft-ietf-tls-esni.md

```diff
@@ -442,14 +454,16 @@ standard ClientHello, with the exception of the following rules:
    incompatible with ECH.
 1. It MUST NOT offer to resume any session for TLS 1.2 and below.
 1. It SHOULD contain TLS padding {{!RFC7685}} as described in {{padding}}.
+1. If it intends to compress any extensions (see {{outer-extensions}}), it
+   MUST order those extensions consecutively.
```

Comment:
It should be possible to relax this for the existing "inner_digest" mechanism. However, I suspect this is will be essential once #336 lands.

Reply:
Hrm, I'm not sure I follow. I was thinking less a security requirement and more a correctness/syntax one. The outer_extensions mechanism, inner digest or AAD, can't express more than one consecutive chunk of slides extensions. That means that when you assemble the ClientHello, you need to adjust the order to fit your intended compression.

Reply:
My point is that the map from ClientHelloInner to EncodedClientHelloInner need not be invertible. To compute "inner_digest", the client just needs to be able to compute the same map from EncodedClientHelloInner to ClientHelloInner that the server does. It's not necessary that the extensions be consecutive, although it would be much easier to implement this correctly if they were.

In any case, I'm happy with this rule.

Reply:
I'm still a little confused. Just double-checking that the text is clear enough and we're on the same page... (If we're both happy with the text, but because we interpreted it differently, that's not very helpful!)

I am saying it *is* important that ClientHelloInner to EncodedClientHelloInner is invertible. This is true in the status quo, and it remains true after #336. Handshake correctness requires that the server reconstruct the *exact* same value, byte-for-byte. Even though the extension order isn't very meaningful semantically, it changes what goes in the transcript and binders, so both sides need to agree. So if the client first assembles:

`ClientHelloInner = (key_share, supported_versions, supported_groups, pre_shared_keys)`

then it compresses key_share + supported_groups as:

`EncodedClientHelloInner = (outer_extensions(key_share, supported_groups), supported_versions, pre_shared_keys)`

The server will reconstruct:

`ClientHelloInner' = (key_share, supported_groups, supported_versions, pre_shared_keys)`

These don't match, so the handshake will fail, both at the transcript and at checking the PSK binders. This isn't a security question because we haven't gotten there yet. It's a basic correctness question.

Now, the thing that went wrong here is that the entire process of swapping extensions out for outer_extensions *only works if the removed extensions are consecutive*. Thus, this PR adds "Removed extensions MUST be ordered consecutively in ClientHelloInner" to EncodedClientHelloInner construction. For that precondition to hold, the client must have ordered the extensions it wanted to compress, thus we add a matching rule, this sentence, to ClientHelloInner construction. Now everything works out and the compression process is correct.

Reply:
Oh, I think I might see the misunderstanding.

> [...] the client just needs to be able to compute the same map from EncodedClientHelloInner to ClientHelloInner that the server does.

This does not work. The client does not compute EncodedClientHelloInner to ClientHelloInner at all. It cannot. By the time we've gotten this far, we've already generated all the PSK binders.

Reply:
Oof, I forgot about the PSK binders :D Thanks for taking the time to clarify my misunderstanding.